Firestore/Source/Util/FSTDispatchQueue.mm

```diff
@@ -145,20 +147,20 @@ - (void)markDone {
 
 #pragma mark - FSTDispatchQueue
 
-@interface FSTDispatchQueue ()
+@interface FSTDispatchQueue () {
+  /**
+   * Flag set while an FSTDispatchQueue operation is currently executing. Used for assertion
+   * sanity-checks.
+   */
+  std::atomic<bool> _operationInProgress;
+}
 
 /**
  * Callbacks scheduled to be queued in the future. Callbacks are automatically removed after they
  * are run or canceled.
  */
 @property(nonatomic, strong, readonly) NSMutableArray<FSTDelayedCallback *> *delayedCallbacks;
 
-/**
- * Flag set while an FSTDispatchQueue operation is currently executing. Used for assertion
- * sanity-checks.
- */
-@property(nonatomic, assign) BOOL operationInProgress;
```

Comment:
If I change the property to be `atomic`, TSan still reports a data race. I probably don't understand how Objective-C `atomic` works.

Reply:
Objective-C properties are a code-generation mechanism. The attributes here manipulate the generated code for the getter and setter. In this case, the value is being accessed through the instance variable directly so the change has no effect.

If you changed the references from `_operationInProgress` to `self.operationInProgress` (and flipped the atomic attribute) that should also fix this.